ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/EarliestByOffsetFactory.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2022 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.function.udaf.offset;
+
+import com.google.common.collect.ImmutableList;
+import io.confluent.ksql.function.AggregateFunctionFactory;
+import io.confluent.ksql.function.AggregateFunctionInitArguments;
+import io.confluent.ksql.function.KsqlAggregateFunction;
+import io.confluent.ksql.function.types.ParamType;
+import io.confluent.ksql.schema.ksql.SchemaConverters;
+import io.confluent.ksql.schema.ksql.SqlArgument;
+import io.confluent.ksql.schema.ksql.types.SqlArray;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.util.KsqlException;
+import java.util.List;
+import org.apache.kafka.connect.data.Schema;
+
+public class EarliestByOffsetFactory extends AggregateFunctionFactory {
+  static final String FUNCTION_NAME = "EARLIEST_BY_OFFSET";
+
+  public EarliestByOffsetFactory() {
+    super(FUNCTION_NAME);
+  }
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  @Override
+  public KsqlAggregateFunction createAggregateFunction(
+      final List<SqlArgument> argTypeList,
+      final AggregateFunctionInitArguments initArgs
+  ) {
+    // This parses the various initArgs that could be used.
+    // If there are two arguments, then they should be (Boolean, Integer)
+    // A single parameter could be passed in to handle either setting "n" or "ignoreNulls".
+    boolean ignoreNulls = true;
+    int n = 1;
+
+    if (initArgs != null) {
```

Comment:
Doesn't argTypeList contain the input parameters?

Reply:
Good question.  It may contain the data that's been pulled out into `initArgs`.  If so, maybe some of this mess can be simplified...

Either way, the `argTypeLIst` is used down line 71 to get the type of the input column.